### ISSUE TITLE
[BUG] Invoked macro in spite of faulty syntax

### ISSUE THREAD

**Issue by twMat** [Jermolene/TiddlyWiki5#5749]
**Describe the bug**
It seems a macro is called even if the syntax is faulty.

**To Reproduce**
Put the following in a tiddler and save. Note that the macrocall is missing a closing `>`. 

```
\define mymacro() should not appear

<<mymacro>
<<currentTiddler>>
```
Output:

`should not appear`

I was not expecting the macro to run successfully and suspect this is not the intention either.

**Comment by twMat** [comment]
@pmario - thanks. 
So I'm guessing I should close this issue. Somebody object if I shouldn't.

**Comment by pmario** [comment]
You may reopen it. ... I do have an idea, where to look at. ... Or may be @Jermolene is faster ;)

**Comment by Jermolene** [comment]
I don't think there's any bug here @pmario. The OP is interpreted as a macrocall to "mymacro" with the parameters `>` and `<<currentTiddler`. What other interpretation would you expect?

**Comment by pmario** [comment]
As I wrote in my first post, I also think it's not a bug. .. It's strange behaviour. ... I will have a look at the parser, if it would be possible to create e regexp, that would be able to detect the "strange behaviour"

`<<mymacro>` ... As you can see: `>` is detected as the first parameter, even if there no space between the macro name and `>` which imo _is strange_ behaviour, which should be fixed if we can. 

I would have expected the `>` to be part of the macro name. ...

**Comment by Jermolene** [comment]
> As I wrote in my first post, I also think it's not a bug. .. It's strange behaviour. ... I will have a look at the parser, if it would be possible to create e regexp, that would be able to detect the "strange behaviour"
> 
> `<<mymacro>` ... As you can see: `>` is detected as the first parameter, even if there no space between the macro name and `>` which imo _is strange_ behaviour, which should be fixed if we can.
> 
> I would have expected the `>` to be part of the macro name. ...

It's all swings and roundabouts. The parser limits the acceptable characters within macro names, and doesn't require a space after the macro name. Thus, `<<tag"HelloThere">>` (with no space after the macroname) works as users expect, rather than being interpreted as a call to a macro named `tag"HelloThere"`.

**Comment by pmario** [comment]
You are right. I did have a look at the "macro name regexp": `var reMacroName = /([^\s>"'=]+)/g;`

It means that the following chars END the macro name and start the parameter section: ... Whitespace, `>`,`"`, `'` and `=` 

@twMat ... IMO it's a "strange edge case". ... We may be able to add some info to the doc, that the described chars will end the macro name ans start the macro-parameter section ... 

Or we just say the following chars can not be part of the macro-name, which may be easier to understand: Whitespace, `>`,`"`, `'` and `=` 

But I'm not sure, it this would have helped you?! .

**Comment by pmario** [comment]
>It's all swings and roundabouts.

q:-))  ... Yea, Sometimes it's a miracle, how "forgiving" the TW syntax is. ... Most of the times it "just works" as intended. ... But Mat finds the edge-cases ;)

**Comment by twMat** [comment]
>Or we just say the following chars can not be part of the macro-name, which may be easier to understand: Whitespace, `>,", '` and `=`

Yes, I think it would be appropriate to caution against these special characters in macro names. Not sure people will read it, or know to look it up if they hit this "strange behaviour", but it would make sense in the docs.

I can add this to the docs (in a few weeks... I'm procrastinating some things I *must* do) but just so I don't misunderstand: Those characters - `whitespace`, `>,", '` and `=` are thus all *allowed*, right?, but they can cause this edge case behaviour. Correct?

**Comment by pmario** [comment]
>Thus, <<tag"HelloThere">> (with no space after the macroname) works as users expect, rather than being interpreted as a call to a macro named tag"HelloThere".

@Jermolene ... This doesn't work on with 5.1.23 ... but it works with the prerelease ... That's strange too

**Comment by saqimtiaz** [comment]
@pmario probably related to #5451

**Comment by twMat** [comment]
Reopening until dust settles.

**Comment by pmario** [comment]
@flibbles Can you have a look

**Comment by flibbles** [comment]
Yeah. This is because of my changes in #5451. The wikirule macrocallinline and macrocallblock were using their own inferior parsers. Meanwhile, the attribute parsers (e.g. `<$text text=<<myMacro myArgs>> />`) were using a much more robust parser inside of parseutils.js. My change made them use the same robust one.

I had to make a few judgement calls along the way, because their behavior wasn't exactly the same, and I see that the <<tag"HelloThere">> and also <<macroWith>karat>> were both edge cases where I made such calls. There are tests regarding those, if I remember.

**Comment by pmario** [comment]
@flibbles Thx for having a look.

The main problem I see with the OT edge-case is, that it "seems" as if the `mymacro` works as intended, but it doesn't. It actually cannibalises the following text until it finds 2 closing brackets at: `<<currentTiddler>>`. It turns `<<currentTiddler` into a parameter and uses the `>>` to close itself.

```
\define mymacro() should not appear

<<mymacro>
<<currentTiddler>>
```

The problem with this behaviour is, that the user will start to search "What's wrong with `<<currentTiddler>>` " and not at the actual problem. ... 

So I personally would prefer that `<<mymacro>` would be transfered into a macro-name. In our example it would show nothing, since the macro definition is different to the macro-name. ... So the user will hopefully start to debug at the right point in the text. 

----

The funny thing is, that all other chars `=`, `"` and `'` are converted to macro names. Only `>` is converted to a parameter. ... That's also 1 thing, that I don't understand, because the regexp is: `var reMacroName = /([^\s>"'=]+)/g;` ... Which says, that all of them should be converted into params. 

But I'm probably missing something in the code. .. I'd need to single-step through it, but I didn't have time to do this.
